julia/core.py

```diff
@@ -40,70 +41,132 @@ def iteritems(d): return iter(d.items())
 else:
     iteritems = dict.iteritems
 
+""" Original environment variables; Used in tests. """
+_orig_env = os.environ.copy()
 
 class JuliaError(Exception):
     pass
 
 
+def remove_prefix(string, prefix):
+    return string[len(prefix):] if string.startswith(prefix) else string
+
+
+def jl_name(name):
+    if name.endswith('_b'):
+        return name[:-2] + '!'
+    return name
+
+
+def py_name(name):
+    if name.endswith('!'):
+        return name[:-1] + '_b'
+    return name
+
+
 class JuliaModule(ModuleType):
-    pass
 
+    def __init__(self, loader, *args, **kwargs):
+        super(JuliaModule, self).__init__(*args, **kwargs)
+        self._julia = loader.julia
+        self.__loader__ = loader
+
+    @property
+    def __all__(self):
+        juliapath = remove_prefix(self.__name__, "julia.")
+        names = set(self._julia.eval("names({})".format(juliapath)))
+        names.discard(juliapath.rsplit('.', 1)[-1])
+        return [py_name(n) for n in names if is_accessible_name(n)]
+
+    def __dir__(self):
+        if python_version.major == 2:
+            names = set()
+        else:
+            names = set(super(JuliaModule, self).__dir__())
+        names.update(self.__all__)
+        return list(names)
+    # Override __dir__ method so that completing member names work
+    # well in Python REPLs like IPython.
+
+    def __getattr__(self, name):
+        try:
+            return self.__try_getattr(name)
+        except AttributeError:
+            if name.endswith("_b"):
+                try:
+                    return self.__try_getattr(jl_name(name))
+                except AttributeError:
+                    pass
+            raise
+
+    def __try_getattr(self, name):
+        juliapath = remove_prefix(self.__name__, "julia.")
+        try:
```

Comment:
instead of this `try` block, maybe we should try calling `isdefined(module, :name)`, throwing an `AttributeError` if it returns `false`, and letting any `JuliaError` that is thrown fall through to the caller.